src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,766 @@
+
+        // to avoid convoying, each silo randomizes these period intervals
+        private readonly TimeSpan randomizedBackgroundGossipInterval;
+        private TimeSpan randomizedResendActiveStatusAfter;
+
+        private GrainTimer timer;
+        private ISiloStatusOracle siloStatusOracle;
+        private MultiClusterConfiguration injectedConfig;
+
+        public MultiClusterOracle(SiloAddress silo, List<IGossipChannel> sources, GlobalConfiguration config)
+            : base(Constants.MultiClusterOracleId, silo)
+        {
+            if (sources == null) throw new ArgumentNullException("sources");
+            if (silo == null) throw new ArgumentNullException("silo");
+
+            logger = TraceLogger.GetLogger("MultiClusterOracle");
+            gossipChannels = sources;
+            localData = new MultiClusterOracleData(logger);
+            clusterId = config.ClusterId;
+            defaultMultiCluster = config.DefaultMultiCluster;  
+            random = new SafeRandom();
+            randomizedBackgroundGossipInterval = RandomizeTimespan(config.BackgroundGossipInterval);
+            randomizedResendActiveStatusAfter = RandomizeTimespan(ResendActiveStatusAfter);
+        }
+   
+        // randomize a timespan by up to 10%
+        private TimeSpan RandomizeTimespan(TimeSpan value)
+        {
+            return TimeSpan.FromMilliseconds(value.TotalMilliseconds * (0.9 + (random.NextDouble() * 0.1)));
+        }
+
+        public bool IsFunctionalClusterGateway(SiloAddress siloAddress)
+        {
+            GatewayEntry g;
+            return localData.Current.Gateways.TryGetValue(siloAddress, out g) 
+                && g.Status == GatewayStatus.Active;
+        }
+
+        public IEnumerable<string> GetActiveClusters()
+        {
+            var clusters = localData.Current.Gateways.Values
+                 .Where(g => g.Status == GatewayStatus.Active)
+                 .Select(g => g.ClusterId)
+                 .Distinct();
+
+            return clusters;
+        }
+
+        public IEnumerable<GatewayEntry> GetGateways()
+        {
+            return localData.Current.Gateways.Values;
+        }
+
+        public SiloAddress GetRandomClusterGateway(string cluster)
+        {
+            var activeGateways = this.localData.Current.Gateways.Values
+                .Where(gw => gw.ClusterId == cluster && gw.Status == GatewayStatus.Active)
+                .Select(gw => gw.SiloAddress)
+                .ToList();
+
+            if (activeGateways.Count == 0)
+                return null;
+
+            return activeGateways[random.Next(activeGateways.Count)];
+        }
+
+        public MultiClusterConfiguration GetMultiClusterConfiguration()
+        {
+            return localData.Current.Configuration;
+        }
+
+        public async Task InjectMultiClusterConfiguration(MultiClusterConfiguration config)
+        {
+            this.injectedConfig = config;
+            PushChanges();
```

Comment:
might want to log an Informational message about this, or is this covered somewhere else?